src/synchronizer.rs

```diff
@@ -236,13 +246,103 @@ impl<H: Hasher + Default, const N: usize, const SD: u64> Synchronizer<H, N, SD>
     /// whether synchronized `entity` has changed.
     pub fn version(&mut self) -> Result<InstanceVersion, SynchronizerError> {
         // fetch current state from mapped memory
-        let state = self.state_container.state(false)?;
+        let state = self.state_container.state_read(false)?;
 
         // fetch current version
         state.version()
     }
 }
 
+/// Configures how a synchronizer uses write locks.
+#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
+pub enum WriteLockMode {
+    /// Write-locking is disabled. This mode implies that there is only a single writer or that
+    /// writers are synchronized out-of-band.
+    #[default]
+    Disabled,
+    /// Use locks to ensure that only a single register is registered. The lock is registered once
+    /// by the writer and held until the writer terminates.
+    SingleWriter,
+}
+
+/// A builder to configure [`Synchronizer`] options.
+#[derive(Clone, Debug)]
+pub struct SynchronizerBuilder<'a> {
+    path_prefix: &'a OsStr,
+    write_lock_mode: WriteLockMode,
+}
+
+impl<'a> SynchronizerBuilder<'a> {
+    /// Create a new builder using the provided path prefix.
+    pub fn new(path_prefix: &'a OsStr) -> Self {
+        Self {
+            path_prefix,
+            write_lock_mode: WriteLockMode::Disabled,
```

Comment:
Thinking out loud - what if default write mode was `SingleWriter` - since if a caller is building/user a Synchronizer it seems the default behaviour should be locking rather than always explicitly enabling `SingleWriter`?

Reply:
Can you clarify what you mean? Are you suggesting that `WriteLockMode::SingleWriter` should be the default at all times (i.e., that we forego the `write-lock` feature flag)? Or that it should become the default automatically when the `write-lock` feature is enabled?

There are a few reasons this functionality is not the default and is gated behind a feature flag:
1. Performance. We want to maintain [zero-cost abstractions](https://blog.rust-lang.org/2015/05/11/traits.html), or in Stroustrup's words "What you don't use, you don't pay for." @bocharov optimized writer performance in #11; by placing this code behind a feature flag we avoid performance regression for applications that do not require write locking.
2. Compatibility. `flock` is only available on Unix systems, so the `libc` dependency and all calls to it must be gated behind a feature flag to retain the Windows support added in #3. (In theory, we may be able to add some Windows-compatible locking mechanism in the future; but I'm unlikely to spend any time on this.)

Perhaps we could experiment further with the configuration API. E.g., define a `WriteLockStrategy` trait to support this. That would become an additional generic parameter on `Synchronizer`, allowing us to skip the builder pattern. Still, I wouldn't make `SingleWriter` a default; IMO, toggling the feature flag shouldn't change the configuration of any `Synchronizer` without explicitly opting in.

Reply:
I'm not talking about the feature flag. "that it should become the default automatically when the write-lock feature is enabled?" - yes.

Reply:
That carries the risk of changing the behavior of existing code when the flag is enabled/disabled. Typically, feature flags enable/expose new APIs rather than controlling behavior directly. This is best as it should be possible to mix locking and non-locking `Synchronizers` even within a single project.

Reply:
I guess I don't really see the distinction between enabling (and ostensibly using) new APIs rather than controlling behaviour. Point being is - what is the use for a `Synchronizer` if not locking (ignoring feature flags for the sake of argument)?